File: 백준/Silver/2156. 포도주 시식/포도주 시식.c
#define _CRT_SECURE_NO_WARNINGS
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

#include<stdio.h>
int main() {
	int a = 0;
	scanf("%d", &a);
	int arr[10001] = { 0, };
	int arr2[3][10001] = { 0, };
	for (int i = 0; i < a; i++) {
		scanf("%d", &arr[i]);
	}
	int max = 0;
	if (a == 1)
		printf("%d", arr[0]);
	else if (a == 2)
		printf("%d", arr[0] + arr[1]);
	else {
		arr2[0][1] = arr[0];
		arr2[1][1] = arr[1];
		arr2[2][1] = arr[0] + arr[1];


		for (int j = 2; j < a; j++) {
			arr2[0][j] = MAX(arr2[2][j - 1], arr2[1][j - 1]);
			arr2[0][j] = MAX(arr2[0][j], arr2[0][j - 1]);
			arr2[1][j] = arr[j] + arr2[0][j - 1];
			arr2[2][j] = arr[j] + arr2[1][j - 1];
		}
		max = MAX(arr2[0][a - 1], arr2[1][a - 1]);
		max = MAX(max, arr2[2][a - 1]);
		printf("%d", max);
			
	}
}

File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
n= int(input())
arr = []
dp = [0]*n
for _ in range(n):
    arr.append(int(input()))

dp[0] = arr[0]
if n >= 2:
    dp[1] = arr[0]+arr[1]
if n >= 3:
    dp[2] = max(arr[0]+arr[1],arr[0]+arr[2],arr[1]+arr[2])
for i in range(3,n):
    dp[i] = max(arr[i] + dp[i-2], arr[i] + arr[i-1]+ dp[i-3], dp[i-1])

print(dp[n-1])